buku

```diff
@@ -599,21 +623,45 @@ class BukuDb:
         rows = self._fetch(query + ' LIMIT 1', *args, lock=lock)
         return rows[0] if rows else None
 
-    def get_rec_all(self, *, lock: bool = True):
+    def _ordering(self, fields=['+id'], for_db=True) -> List[Tuple[str, bool]]:
+        """Converts field list to ordering parameters (for DB query or entity list sorting).
+        Fields are listed in priority order, with '+'/'-' prefix signifying ASC/DESC; assuming ASC if not specified.
+        Other than names from DB, you can pass those from JSON export."""
+        names = {'index': 'id', 'uri': 'url', 'description': 'desc', **({'title': 'metadata'} if for_db else {'metadata': 'title'})}
+        valid = list(names) + list(names.values()) + ['tags']
+        _fields = [(re.sub(r'^[+-]', '', s), not s.startswith('-')) for s in (fields or [])]
+        _fields = [(names.get(field, field), direction) for field, direction in _fields if field in valid]
+        return _fields or [('id', True)]
+
+    def _sort(self, records: List[BookmarkVar], fields=['+id'], ignore_case=True) -> List[BookmarkVar]:
+        text_fields = (set() if not ignore_case else {'url', 'desc', 'title', 'tags'})
+        get = lambda x, k: (getattr(x, k) if not ignore_case or k not in text_fields else str(getattr(x, k) or '').lower())
+        order = self._ordering(fields, for_db=False)
+        return sorted(bookmark_vars(records), key=lambda x: [SortKey(get(x, k), ascending=asc) for k, asc in order])
```

Comment:
Sorting already fetched data… is not actually used anywhere in the code, but it might come in handy anytime.